app/_src/kubernetes-ingress-controller/guides/security/verify-upstream-tls.md

```diff
@@ -0,0 +1,392 @@
+service accordingly.
+
+```shell
+kubectl annotate service echo konghq.com/tls-verify=true
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see an error similar to this.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+{
+  "message":"An invalid response was received from the upstream server",
+  "request_id":"e2b3182856c96c23d61e880d0a28012f"
+}
+```
+
+If you inspect {{ site.base_gateway }}'s container logs, you should see an error indicating an issue with TLS handshake.
+
+```shell
+ kubectl logs -n kong deploy/kong-gateway | grep "GET /echo"
+```
+
+```text
+2024/11/28 18:04:05 [error] 1280#0: *20004 upstream SSL certificate verify error: (20:unable to get local issuer certificate) while SSL handshaking to upstream, client: 192.168.194.1, server: kong, request: "GET /echo HTTP/1.1", upstream: "https://192.168.194.9:443/", host: "kong.example", request_id: "e2b3182856c96c23d61e880d0a28012f"
+192.168.194.1 - - [28/Nov/2024:18:04:05 +0000] "GET /echo HTTP/1.1" 502 126 "-" "curl/8.7.1" kong_request_id: "e2b3182856c96c23d61e880d0a28012f"
+```
+
+{{ site.base_gateway }} is now verifying the certificate of the upstream service and rejecting the connection because
+the certificate is not trusted.
+
+### Add the root CA certificate to {{ site.base_gateway }}
+
+That can be fixed by adding the root CA certificate to the {{ site.base_gateway }}'s CA certificates and associating
+it with the service.
+
+First, create a secret with the root CA certificate.
+
+{% include /md/kic/ca-certificates-note.md %}
+
+```shell
+echo '
+apiVersion: v1
+kind: Secret
+metadata:
+  name: root-ca
+  labels:
+    konghq.com/ca-cert: "true" # This label is required for the CA certificate to be recognized by Kong
+  annotations:
+    kubernetes.io/ingress.class: kong
+data:
+  cert: '$(base64 -w0 ./certs/root.crt)'
+  # An arbitrary ID for the certificate 
+  id: '$(printf "bf6e0f14-78cd-45ad-9325-87ec7ef7b890" | base64 -w0)'
+' | kubectl apply -f -
+```
+
+The results should look like this.
+
+```text
+secret/root-ca created
+```
+
+Now, associate the root CA certificate with the service passing its name to `konghq.com/ca-certificates` annotation.
+
+{:.note}
+> The `konghq.com/ca-certificates` annotation is a comma-separated list of CA certificate names. You can add multiple
+> CA certificates to the list.
+
+```shell
+kubectl annotate service echo konghq.com/ca-certificates='root-ca'
+```
+
+The results should look like this.
+
+```text
+service/echo annotated
+```
+
+Now, when you issue the same request as before, you should see a successful response.
+
+```shell
+curl -H Host:kong.example $PROXY_IP/echo
+```
+
+```text
+Welcome, you are connected to node orbstack.
+Running on Pod echo-bd94b7dcc-gdnhl.
+In namespace default.
+With IP address 192.168.194.18.
+Through HTTPS connection.
+```
```

Comment:
At this point I'm getting 

```
curl -H "Host: kong.example" $PROXY/echo
{
  "message":"An invalid response was received from the upstream server",
  "request_id":"e84c7429cc4714117be1956514ec8142"
}
```

Because we've changed the `konghq.com/ca-certificates` annotation into 2:

- `konghq.com/ca-certificates-secret`
- `konghq.com/ca-certificates-configmap`

Allowing configmaps is dictated by the Gateway API spec in: https://github.com/kubernetes-sigs/gateway-api/blob/26051f587ccaf54823ccfd7eca5cbf2013aa8177/apis/v1alpha3/backendtlspolicy_types.go#L111-L116

I'll update the guide so that it reflects these changes.

Reply:
8283d58d647d9607e0de8f798fdea81264638613 contains the changes.

Reply:
Checked and it works!